src/storage-client/src/controller/remap_migration.rs

```diff
@@ -0,0 +1,358 @@
+        write_shard: ShardId,
+        // additional information to improve logging
+        id: GlobalId,
+        partitioned_source: bool,
+    ) -> anyhow::Result<Option<Self>> {
+        let mut persist_clients = persist_clients.lock().await;
+        let persist_client = persist_clients
+            .open(read_metadata.persist_location.clone())
+            .await
+            .context("error creating persist client")?;
+        drop(persist_clients);
+
+        let (current_shard_write_handle, current_shard_read_handle) = persist_client
+            .open::<crate::types::sources::data::SourceData, (), T, Diff>(
+                read_metadata.remap_shard,
+                &format!("reclock {}", id),
+            )
+            .await
+            .context("error opening persist shard")?;
+
+        let (since, upper) = (
+            current_shard_read_handle.since(),
+            current_shard_write_handle.upper().clone(),
+        );
+
+        let as_of = match upper.as_option() {
+            Some(u) if u > &T::minimum() => Antichain::from_elem(u.step_back().unwrap()),
+            // Source is terminated or new; nothing to migrate
+            _ => return Ok(None),
+        };
+
+        assert!(
+            PartialOrder::less_equal(since, &as_of),
+            "invalid as_of: as_of({as_of:?}) < since({since:?}), \
+            source {id}, \
+            remap_shard: {}",
+            read_metadata.remap_shard
+        );
+
+        assert!(
+            as_of.elements() == [T::minimum()] || PartialOrder::less_than(&as_of, &upper),
+            "invalid as_of: upper({upper:?}) <= as_of({as_of:?})",
+        );
+
+        tracing::info!(
+            ?since,
+            ?as_of,
+            ?upper,
+            "RemapHandleMigrator({id}) initializing"
+        );
+
+        Ok(Some(Self {
+            persist_client,
+            current_shard_read_handle,
+            as_of,
+            write_shard,
+            partitioned_source,
+        }))
+    }
+}
+
+/// Unpacks a Kafka remap binding from a Row. Returns `None` if encounters the
+/// new format, in which case the migration doesn't need to go any further.
+///
+/// # Panics
+/// If called on data that would not be generated in either the legacy or new
+/// schema of a Kafka source's remap shard.
+fn unpack_kafka_remap_binding(data: SourceData) -> Option<(PartitionId, MzOffset)> {
+    let row = data.0.expect("invalid binding");
+    let mut datums = row.iter();
+    match (datums.next(), datums.next()) {
+        // Legacy binding
+        (Some(Datum::Int32(p)), Some(Datum::UInt64(offset))) => {
+            Some((PartitionId::Kafka(p), MzOffset::from(offset)))
+        }
+        // Current binding
+        (Some(Datum::Range(_)), Some(Datum::UInt64(_))) => None,
+        _ => panic!("invalid Kafka remap data in collection"),
+    }
+}
+
+#[async_trait::async_trait(?Send)]
+impl<T: Timestamp + Lattice + TotalOrder + Codec64 + From<EpochMillis> + TimestampManipulation>
```

Comment:
We shouldn't need `TimestampManipulation` here once we start at `since` and go all the way to `upper`